core/templates/pages/teach-page/teach-page.component.spec.ts

```diff
@@ -15,112 +15,193 @@
+        this._hash = val;
+        if (this._hashChange === null) {
+          return;
+        }
+        this._hashChange();
+      },
+      get href() {
+        return this._href;
+      },
+      set href(val) {
+        this._href = val;
+      },
+      reload: (val) => val
+    },
+    get onhashchange() {
+      return this.location._hashChange;
+    },
+
+    set onhashchange(val) {
+      this.location._hashChange = val;
+    }
+  };
+  get nativeWindow() {
+    return this._window;
+  }
+}
+
+class MockSiteAnalyticsService {
+  registerApplyToTeachWithOppiaEvent(): void {
+    return;
+  }
+}
+
+let component: TeachPageComponent;
+let fixture: ComponentFixture<TeachPageComponent>;
 
-    ctrl = $componentController('teachPage', {
-      $rootScope: $scope
-    });
-  }));
+describe('Teach Page', function() {
+  let windowRef: MockWindowRef;
+  let sas = null;
+  let i18n = null;
+  let translate = null;
+
+  beforeEach(() => {
+    windowRef = new MockWindowRef();
+    TestBed.configureTestingModule({
+      declarations: [TeachPageComponent, MockTranslatePipe],
+      providers: [
+        {
+          provide: I18nLanguageCodeService,
+          useClass: MockI18nLanguageCodeService
+        },
+        { provide: SiteAnalyticsService, useClass: MockSiteAnalyticsService },
+        { provide: TranslateService, useClass: MockTranslateService },
+        UrlInterpolationService,
+        { provide: WindowRef, useValue: windowRef }
+      ],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+    i18n = TestBed.get(I18nLanguageCodeService);
+    translate = TestBed.get(TranslateService);
+    sas = TestBed.get(SiteAnalyticsService);
+    fixture = TestBed.createComponent(TeachPageComponent);
+    component = fixture.componentInstance;
+  });
 
-  afterEach(function() {
+  afterEach(() => {
     // Property onhashchange and location.hash are reassigned because it shares
     // same memory reference to all test blocks and the controller itself
     // because $provide.value of WindowRef refers to windowRef as well.
     // Once location.hash or onhashchange is setted in the controller,
     // the value will be only available in the test block itself, not affecting
     // others test block.
+    windowRef.nativeWindow.location.href = '';
     windowRef.nativeWindow.onhashchange = null;
     windowRef.nativeWindow.location.hash = '';
   });
 
-  it('should click on teach tab', function(done) {
-    ctrl.$onInit();
-    expect(ctrl.activeTabName).toBe('teach');
+  it('should click on teach tab', () => {
+    component.ngOnInit();
+    expect(component.activeTabName).toBe('teach');
 
-    ctrl.onTabClick('teach');
+    component.onTabClick('teach');
 
-    // Function setTimeout is being used here in order to wait onhashchange
-    // event to finish. setTimeout is executed only after call stack is empty.
-    // Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop
-    setTimeout(function() {
```

Comment:
This is removed because we are using mocks and it is not taking any time to reflect the changes, right?

Reply:
The mocks don't take time to reflect the changes, but their is a setTimeout in the actual function call. I don't know what was the original motivation to add that, I think just to make the transition smoother? I was getting a race condition with setTimeout in the tests as well, and they were very flaky. I'll have another shot with using tick and trying setTimeout instead.

Reply:
*without using tick

Reply:
Have yo tried?

Reply:
It's already done in the recent commits